x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/idp/SamlMetadataGenerator.java

```diff
@@ -0,0 +1,98 @@
+
+    public SamlMetadataGenerator(SamlFactory samlFactory, SamlIdentityProvider idp) {
+        this.samlFactory = samlFactory;
+        this.idp = idp;
+        SamlInit.initialize();
+    }
+
+    public void generateMetadata(String spEntityId, ActionListener<SamlGenerateMetadataResponse> listener) {
+        try {
+            // this will be async
+            SamlServiceProvider sp = idp.getRegisteredServiceProvider(spEntityId);
+            if (null == sp) {
+                listener.onFailure(new IllegalArgumentException("Service provider with Entity ID [" + spEntityId
+                    + "] is not registered with this Identity Provider"));
+                return;
+            }
+            EntityDescriptor metadata = buildEntityDescriptor(sp);
+            final X509Credential signingCredential = idp.getMetadataSigningCredential();
+            Element metadataElement = possiblySignDescriptor(metadata, signingCredential);
+            listener.onResponse(new SamlGenerateMetadataResponse(samlFactory.toString(metadataElement, false)));
+        } catch (Exception e) {
+            logger.debug("Error generating IDP metadata to share with [" + spEntityId + "]", e);
+            listener.onFailure(e);
+        }
+    }
+
+    EntityDescriptor buildEntityDescriptor(SamlServiceProvider sp) throws Exception {
+        final SamlIdPMetadataBuilder builder = new SamlIdPMetadataBuilder(idp.getEntityId())
+            .wantAuthnRequestsSigned(sp.shouldSignAuthnRequests())
+            .withSingleSignOnServiceUrl(SAML2_REDIRECT_BINDING_URI,
+                idp.getSingleSignOnEndpoint(SAML2_REDIRECT_BINDING_URI))
+            .withSingleSignOnServiceUrl(SAML2_POST_BINDING_URI,
+                idp.getSingleSignOnEndpoint(SAML2_POST_BINDING_URI))
+            .withSingleLogoutServiceUrl(SAML2_REDIRECT_BINDING_URI,
+                idp.getSingleLogoutEndpoint(SAML2_REDIRECT_BINDING_URI))
+            .withSingleLogoutServiceUrl(SAML2_POST_BINDING_URI,
+                idp.getSingleLogoutEndpoint(SAML2_POST_BINDING_URI))
+            .withSigningCertificate(idp.getSigningCredential().getEntityCertificate())
+            .withNameIdFormat(PERSISTENT)
+            .withNameIdFormat(TRANSIENT)
+            .organization(idp.getOrganization())
+            .withContact(idp.getTechnicalContact());
+        return builder.build();
+    }
+
+    Element possiblySignDescriptor(EntityDescriptor descriptor, X509Credential signingCredential) throws MarshallingException,
+        SignatureException {
+        EntityDescriptorMarshaller marshaller = new EntityDescriptorMarshaller();
+        if (null == signingCredential) {
+            return marshaller.marshall(descriptor);
+        } else {
+            Signature signature = samlFactory.buildObject(Signature.class, DEFAULT_ELEMENT_NAME);
+            signature.setSigningCredential(signingCredential);
+            signature.setSignatureAlgorithm(ALGO_ID_SIGNATURE_RSA_SHA256);
```

Comment:
Should we make this configurable?

Reply:
Not now. We're already implementing more than the absolute minimum we need. We really need to focus on MVP here.